File: app/src/main/java/com/arif/notifforwarder/PersistentService.java
package com.arif.notifforwarder;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.Service;
import android.content.Intent;
import android.content.IntentFilter;
import android.net.ConnectivityManager;
import android.os.IBinder;

public class PersistentService extends Service {

    public static final String CHANNEL_ID = "nf_ch";
    public static final int    NOTIF_ID   = 1001;

    private final NetworkChangeReceiver netReceiver = new NetworkChangeReceiver();

    @Override
    public void onCreate() {
        super.onCreate();
        createChannel();
        startForeground(NOTIF_ID, buildNotif());
        registerReceiver(netReceiver,
                new IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION));
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        return START_STICKY;
    }

    @Override
    public IBinder onBind(Intent intent) { return null; }

    @Override
    public void onTaskRemoved(Intent rootIntent) {
        startService(new Intent(getApplicationContext(), PersistentService.class));
        super.onTaskRemoved(rootIntent);
    }

    @Override
    public void onDestroy() {
        try { unregisterReceiver(netReceiver); } catch (Exception ignored) {}
        super.onDestroy();
    }

    private void createChannel() {
        NotificationChannel ch = new NotificationChannel(
                CHANNEL_ID, "Notif Forwarder", NotificationManager.IMPORTANCE_MIN);
        ch.setShowBadge(false);
        ch.setSound(null, null);
        ch.enableVibration(false);
        ((NotificationManager) getSystemService(NOTIFICATION_SERVICE))
                .createNotificationChannel(ch);
    }

    private Notification buildNotif() {
        return new Notification.Builder(this, CHANNEL_ID)
                .setContentTitle("Notif Forwarder aktif")
                .setContentText("Meneruskan notifikasi ke Telegram")
                .setSmallIcon(android.R.drawable.ic_popup_sync)
                .setPriority(Notification.PRIORITY_MIN)
                .build();
    }
}

File: app/src/main/java/com/notifforwarder/app/PersistentService.kt
package com.notifforwarder.app

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.Service
import android.content.Intent
import android.os.IBinder

class PersistentService : Service() {

    companion object {
        const val CHANNEL_ID = "notif_forwarder_channel"
        const val NOTIF_ID = 1001
    }

    override fun onCreate() {
        super.onCreate()
        createNotificationChannel()
        startForeground(NOTIF_ID, buildNotification())
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        // START_STICKY = sistem akan restart service ini jika dibunuh
        return START_STICKY
    }

    override fun onBind(intent: Intent?): IBinder? = null

    override fun onTaskRemoved(rootIntent: Intent?) {
        // Restart diri sendiri jika app di-swipe dari recent
        val restartIntent = Intent(applicationContext, PersistentService::class.java)
        startService(restartIntent)
        super.onTaskRemoved(rootIntent)
    }

    private fun createNotificationChannel() {
        val channel = NotificationChannel(
            CHANNEL_ID,
            "Notif Forwarder",
            // IMPORTANCE_MIN = tidak ada suara, tidak ada popup, tidak tampil di status bar
            NotificationManager.IMPORTANCE_MIN
        ).apply {
            description = "Layanan penerusan notifikasi"
            setShowBadge(false)
        }
        val manager = getSystemService(NotificationManager::class.java)
        manager.createNotificationChannel(channel)
    }

    private fun buildNotification(): Notification {
        return Notification.Builder(this, CHANNEL_ID)
            .setContentTitle("Notif Forwarder aktif")
            .setContentText("Meneruskan notifikasi ke Telegram")
            // Ikon kecil tidak mencolok (pakai ikon sistem)
            .setSmallIcon(android.R.drawable.ic_dialog_info)
            .setPriority(Notification.PRIORITY_MIN)
            .setSilent(true)
            .build()
    }
}
